src/fun/transform/expand_generated.rs

```diff
@@ -1,24 +1,107 @@
+use std::collections::{BTreeSet, HashMap, HashSet};
+
 use crate::{
-  fun::{Book, Term},
+  fun::{Book, Name, Term},
   maybe_grow,
 };
 
-/// Dereferences any generated definitions in the term.
+/// Dereferences any non recursive generated definitions in the term.
 /// Used after readback.
 impl Term {
-  pub fn expand_generated(&mut self, book: &Book) {
+  pub fn expand_generated(&mut self, book: &Book, recursive_defs: &RecursiveDefs) {
     maybe_grow(|| {
       if let Term::Ref { nam } = &*self {
-        if nam.is_generated() {
+        if nam.is_generated() && !recursive_defs.contains(nam) {
           *self = book.defs.get(nam).unwrap().rule().body.clone();
         }
       }
 
-      // Note: this assumes that there will never be a loop of generated functions.
-      // This is true right now, but not necessarily in the future.
       for child in self.children_mut() {
-        child.expand_generated(book);
+        child.expand_generated(book, recursive_defs);
       }
     })
   }
 }
+
+type DepGraph = HashMap<Name, HashSet<Name>>;
+type Cycles = Vec<Vec<Name>>;
+type RecursiveDefs = BTreeSet<Name>;
+
+pub fn find_recursive_defs(book: &Book) -> RecursiveDefs {
```

Comment:
Can you instead make this 
```
impl Book {
  pub fn recursive_defs(&self) -> RecursiveDefs {
    ...
  }
}
```
To be more in line with how we expose other functions on Book and Term? (like Term::free_vars for example)